### PR TITLE
PP-6346 Stripe transfer records pay transaction id for reconciliation

### PR DESCRIPTION
**Pull request by sfount**
Require a `charge` or `refund` external ID when making a Stripe
transfer so that these can be used to simply and clearly reconcile
transactions with bank payouts.

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferMetadata.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
+package uk.gov.pay.connector.gateway.stripe.request;
+
+import java.util.Map;
+
+public class StripeTransferMetadata {
+    public static final String STRIPE_CHARGE_ID_KEY = "stripe_charge_id";
+    public static final String RECONCILIATION_TRANSACTION_ID_KEY = "reconciliation_transaction_id";
+
+    private final String stripeChargeId;
+    private final String reconciliationTransactionId;
```
Comment (sfount) [inline src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferMetadata.java]: Proposal to change to `govuk_pay_transaction_external_id` to make this clearer to understand outside of this feature context

File: src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferMetadata.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    public Map<String, String> format() {
+        return Map.of(
+                formatMetadataRequestKey(STRIPE_CHARGE_ID_KEY), this.stripeChargeId,
+                formatMetadataRequestKey(GOVUK_PAY_TRANSACTION_EXTERNAL_ID), this.govukPayTransactionExternalId
+        );
+    }
+    
+    private String formatMetadataRequestKey(String key) {
+        return "metadata[" + key + "]";
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferMetadata.java]:
```suggestion
        return format("metadata[%s]", key);
```

File: src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferInRequest.java [hunk 3]
```diff
@@ -21,9 +21,10 @@ private StripeTransferInRequest(
             String stripeChargeId,
             String idempotencyKey,
             String transferGroup,
-            StripeGatewayConfig stripeGatewayConfig
+            StripeGatewayConfig stripeGatewayConfig,
+            String govukPayTransactionExternalId
     ) {
-        super(amount, gatewayAccount, stripeChargeId, idempotencyKey, stripeGatewayConfig);
+        super(amount, gatewayAccount, stripeChargeId, idempotencyKey, stripeGatewayConfig, govukPayTransactionExternalId);
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferInRequest.java:27]: Can we have an assertion here https://github.com/alphagov/pay-connector/blob/5ac4d251a1b7009cfc4f90260565951e8e4556f3/src/test/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferInRequestTest.java#L26 for new metadata field

File: src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferOutRequest.java [hunk 4]
```diff
@@ -14,9 +14,10 @@ private StripeTransferOutRequest(String amount,
                                      GatewayAccountEntity gatewayAccount,
                                      String sourceTransactionId,
                                      String idempotencyKey,
-                                     StripeGatewayConfig stripeGatewayConfig
+                                     StripeGatewayConfig stripeGatewayConfig,
+                                     String govukPayTransactionExternalId
     ) {
-        super(amount, gatewayAccount, sourceTransactionId, idempotencyKey, stripeGatewayConfig);
+        super(amount, gatewayAccount, sourceTransactionId, idempotencyKey, stripeGatewayConfig, govukPayTransactionExternalId);
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferOutRequest.java:20]: similarly an assertion here https://github.com/alphagov/pay-connector/blob/846e580b882d26a9fc64aae1822ddaf4f0d496e4/src/test/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferOutRequestTest.java#L23 for new metadata field

File: src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferMetadata.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    public StripeTransferMetadata(String stripeChargeId, String reconciliationTransactionId) {
+        this.stripeChargeId = stripeChargeId;
+        this.govukPayTransactionExternalId = reconciliationTransactionId;
+    }
+
+    public static StripeTransferMetadata from(Map<String, String> metadata) {
+        return new StripeTransferMetadata(metadata.get(STRIPE_CHARGE_ID_KEY), metadata.get(GOVUK_PAY_TRANSACTION_EXTERNAL_ID));
+    }
+
+    public Map<String, String> format() {
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferMetadata.java]:
```suggestion
    public Map<String, String> getParams-- or something similar ?() {
```

**Review by kbottla — CHANGES_REQUESTED** [review]
Looks good, just a few minor comments

**Comment by sfount** [comment]
@kbottla thanks for the useful comments, addressed them all in https://github.com/alphagov/pay-connector/commit/f6add7a76cb03ce7cafd1fdfb76e2142103f462f

File: src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferOutRequest.java (hunk 4 above)
Comment (sfount) [reply src/main/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferOutRequest.java:20]: Agreed 👍 good catch

**Review by kbottla — APPROVED** [review]

**Pull request merged**